src/realm/object-store/sync/subscribable.hpp

```diff
@@ -0,0 +1,169 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SUBSCRIBABLE_HPP
+#define REALM_SUBSCRIBABLE_HPP
+
+#include <cstdint>
+#include <unordered_map>
+#include <mutex>
+
+namespace realm {
+
+/// Generic subscribable that allows for coarse, manual notifications
+/// from class type T.
+template <class T>
+struct Subscribable {
+    /// Token that identifies an observer.
+    /// Unsubscribes when deconstructed to
+    /// avoid dangling observers.
+    struct Token {
+        Token(Subscribable* subscribable, uint64_t token)
+            : m_subscribable(subscribable)
+            , m_token(token)
+        {
+        }
+        Token(Token&& other) noexcept
+            : m_subscribable(std::move(other.m_subscribable))
+            , m_token(std::move(other.m_token))
+        {
+            other.m_subscribable = nullptr;
+        }
+        Token& operator=(Token&& other) noexcept
+        {
+            m_subscribable = std::move(other.m_subscribable);
+            m_token = std::move(other.m_token);
+            other.m_subscribable = nullptr;
+            return *this;
+        }
+        Token(const Token&) = delete;
+        Token& operator=(const Token&) = delete;
+
+        ~Token()
+        {
+            if (m_subscribable) {
+                m_subscribable->unsubscribe(*this);
+            }
+        }
+
+        uint64_t value() const
+        {
+            return m_token;
+        }
+
+    private:
+        Subscribable* m_subscribable;
+        uint64_t m_token;
+
+        template <class U>
+        friend struct Subscribable;
+    };
+
+    using Observer = std::function<void(const T&)>;
+
+    Subscribable() = default;
+    Subscribable(const Subscribable& other)
+    {
+        if (&other == this) {
+            return;
+        }
+        std::lock_guard<std::mutex> lock(m_mutex);
+        std::lock_guard<std::mutex> other_lock(other.m_mutex);
+        m_subscribers = other.m_subscribers;
+    }
+    Subscribable(Subscribable&& other) noexcept
+    {
+        if (&other == this) {
+            return;
+        }
+        std::lock_guard<std::mutex> lock(m_mutex);
+        std::lock_guard<std::mutex> other_lock(other.m_mutex);
+        m_subscribers = std::move(other.m_subscribers);
+    }
+    Subscribable& operator=(const Subscribable& other)
+    {
+        if (&other == this) {
+            return *this;
+        }
+        std::lock_guard<std::mutex> lock(m_mutex);
+        std::lock_guard<std::mutex> other_lock(other.m_mutex);
```

Comment:
This deadlocks if two threads call operator= in opposite orders on different threads (which admittedly is very unlikely, but it'll make threadsanitizer unhappy). `std::scoped_lock lock(m_mutex, other.m_mutex);` fixes that and is shorter as a bonus.